Patch Sink/Source interface to provide suspendable primitives and sho…

qwwdfsad opened the pull request:
…w an example of async sink and source

qwwdfsad committed:
Patch Sink/Source interface to provide suspendable primitives and show an example of async sink and source

qwwdfsad committed:
Show a composable AIO example

e5l commented on okio/src/jvmMain/kotlin/okio/RawSink.kt:
```diff
@@ -26,6 +26,9 @@ actual interface RawSink : Closeable, Flushable {
   actual override fun flush()
 
+  @Suppress("ACTUAL_WITHOUT_EXPECT")
+  actual suspend fun flushSuspend() = flush()
```
we also missing the `awaitSpace` operation on `RawSink`

qwwdfsad replied:
Please be more constructive -- what such a method is supposed to do, what is impossible to implement without it and so on. 

A demonstration would be extremely helpful.

qwwdfsad replied:
I am especially interested in a contract of such a method -- for now, there is no notion of "underlying space" in _any_ Okio primitive. What does it mean for `Buffer()` to return from `awaitSpace(100500)`?

If it doesn't have a parameter at all (i.e. copied straight from https://gist.github.com/e5l/55db8601d915259b0b338368f29a9770), then I'm not sure how it is different from `flush`

e5l commented on okio/src/jvmMain/kotlin/okio/aio/AsynchronousFileSink.kt:
```diff
@@ -0,0 +1,77 @@
+
+  override fun flush() {
+    runBlocking {
+      flushSuspend()
```
I would use `flushSuspend` implementation, but call `future.get()` instead of `runBlocking`, so we can drop kotlinx.coroutines dependency in the main source set

swankjesse commented on okio/src/commonMain/kotlin/okio/RawSink.kt:
```diff
@@ -51,6 +51,23 @@ expect interface RawSink : Closeable {
+   *   they will know themselves. Everyone else is not expected to use it anyway, ideally we would like to even hide it
+   *   under different name, but alas it breaks composability as well.
+   */
+  suspend fun flushSuspend() /* = flush() */
```
My best idea for this . . . . call `flush()`, but on the IO dispatcher. That way the caller gets the behavior they want and the system continues to follow the rules of suspending.

e5l replied:
For me, it's ok to keep `flush()` or leave it unimplemented: In most of the cases it should delegate to to the parent Sink `flushSuspend()`

swankjesse commented on okio/src/jvmMain/kotlin/okio/aio/AsynchronousFileSink.kt:
```diff
@@ -0,0 +1,77 @@
+ * Demonstration of the implementation of the source backed by asynchronous file sink
+ * without affecting composability of operation.
+ *
+ * The underlying implementation leverages no internal API and only uses unsafe cursor for zero copy
```
NICE

swankjesse commented on okio/src/jvmMain/kotlin/okio/aio/AsynchronousFileSink.kt:
```diff
@@ -0,0 +1,77 @@
+    // In real world we could've auto-flushed asynchornously and rendezvous with flushSuspend if necessary
+  }
+
+  override suspend fun flushSuspend() {
```
Sweet

swankjesse commented on okio/src/jvmMain/kotlin/okio/aio/AsynchronousFileSource.kt:
```diff
@@ -0,0 +1,130 @@
+      continuation = it
+      val cursor = temporaryBuffer.readAndWriteUnsafe()
+      this.cursor = cursor
+      // TODO: properly work with multiple segment and sizes, not the point of demonstration
```
👍🏻

swankjesse reviewed (commented):
This is rad!

One annoying gotcha that I think harms the demonstration if you wanted to use it for benchmarking etc . . . it’s my understanding that FileChannels don’t actually do async behind the scenes? Only SocketChannel does? Worth checking into the state of the JDK.

e5l commented on okio/src/commonMain/kotlin/okio/RawSource.kt:
```diff
@@ -72,4 +75,30 @@ interface RawSource : Closeable {
+   *     * It is perfectly understandable from an implementation standpoint,
+   *       but may be a bit weird for users.
+   */
+  suspend fun awaitAvailable(atLeastBytes: Long = -1L): Long {
```
We can delegate this method to `read` with a zero capacity buffer by default.

We also don't need any numbers in the result: just a flag if the source is exhausted.

qwwdfsad replied:
But read is supposed to be non-blocking and non-suspending, could you please elaborate on how this is the reasonable default behaviour for an arbitrary source?

Also, it requires changing `read` contract which now can read as many bytes as it wants in its internal structures and give to `read`'s `buffer` only requested number of bytes

qwwdfsad replied:
The number of bytes in the result may (?) be important in order to know how many bytes is allowed to be read via `read` from a purely non-blocking context, in order to e.g. parse part of bytes and wait for the rest of it.

This is exactly the scenario from the second commit:

* Consider a protocol that wants to read length-prefixed UTF-8 string
* It reads length, it calls `awaitAvailable(length)`, but the stream is malformed and returns fewer bytes than length

Three options here:
* Use "boolean flag", invoke "readUtf8StringStrict", fail because there is no `\n`
* Parse byte-by-byte manually and decode it, throwing where necessary. It neglects the library experience and makes authors to do the same work all over again
* Check returned number of bytes and throw a proper exception, potentially reading leftover bytes via separate byte-by-byte codepath for a better error message

swankjesse replied:
Another option: use the same signature as `read`, but make it suspending.

```
suspend fun readSuspending(buffer: Buffer, byteCount: Long): Long
```

Then the buffering class, `Source`, can use this to build other primitives on top: `requestSuspending` and `requireSuspending`, and maybe even `readUtf8LineSuspending`.

The benefit is it's structurally simple. The enormous drawback is it leads us to building everything 2x, and callers now need to choose everywhere.

qwwdfsad replied:
It's indeed structurally much simpler, but I'm not sure if the `suspend` overhead for protocol parsing is reasonable enough to advocate for such approach as the default one

qwwdfsad commented:
@swankjesse `AsyncFileChannels` is indeed not the best thing to use (in any scenario), I was using it mostly for the demonstration because it was easier than running `socket.select` loop :)

It is indeed very slow; it is "truly" asynchronous under the hood on Windows only and is using [overlapped IO](https://docs.microsoft.com/en-us/windows/win32/sync/synchronization-and-overlapped-input-and-output), but it's not making AFS API any better

qwwdfsad committed:
New conceptual model of awaitAvailable

### Motivation

The previous approach had multiple critical flaws

* It attempted to make a concurrency "transparent" to each individual source in hopes that default implementation will be enough to everyone
* It led to composability issues, i.e. 'gzip().buffer().readUtf8Strict()' has been failing **non-deterministically**
* Default implementation implied that due to non-explicit opt-in into asyncrhonously, any code may either throw really weird exception or do a blockign call where it was not supposed to
* Default implementation implies inefficiency: either unexpected blocking call or dependency on kotlinx.coroutines AND Dispatchers.IO

### New approach

* Make asynchronous behaviour **optional** and fail-fast when the API is not ready to be used in asynchronous manner. It saves users from mis-using the API that was not intended to be used asynchronously
* Introduce AwaitPredicate with a straightforward contract that makes composability explicit concern and allows to compose requirements separately. E.g. in 'source.gzip().buffer().awaitContent(AwaitPredicate.Utf8String)', the implementation of 'buffer' awaits data until there is a newline separator, while 'gzip' operates on a different level and awaits until the full gzip-block is read
* Optional behaviour along with predicates also allows API authors to provide a transparent extensions with different concurrent behaviour to avoid state-maching wrangling, e.g. 'gzip()', 'gzipAsync()' and 'gzipMixed'

qwwdfsad commented on okio/src/jvmMain/kotlin/okio/aio/AsyncGzipSource.kt:
```diff
@@ -0,0 +1,180 @@
+import java.util.zip.*
+import okio.*
+
+class AsyncGzipSource(source: RawSource) : RawSource {
```
FYI: it is a copy-pasted version of `GzipSource` where I changed `read` method and `awaitContent`.

Making Gzip full async is much easier than supporting a mixed mode :)

swankjesse commented on okio/src/commonMain/kotlin/okio/RawSource.kt:
```diff
@@ -77,28 +79,45 @@ interface RawSource : Closeable {
+
+    public val AnyBytes: AwaitPredicate = AwaitPredicate { _, _ -> true }
+
+    public val Utf8String: AwaitPredicate =
```
Utf8Line

swankjesse replied:
Maybe a public function like `containsByteString()` and that can be used to implement searching for `\n`, but also things like `\r\n\r\n` to terminate HTTP/1 headers
